File: csharp-api/test/Test/SharedData/test_interop.lua
-- Set some variables that C# can read
local my_obj = sdk.get_managed_singleton("app.PlayerManager")
log.info("my_obj type: " .. my_obj:get_type_definition():get_full_name())
sdk.shared_data.set_variable("app", my_obj)
sdk.shared_data.set_variable("pi", 3.14159)
sdk.shared_data.set_variable("count", 42)
sdk.shared_data.set_variable("stringVal", "Xin chào thế giới!")

re.on_draw_ui(function()
    -- Read what C# wrote
    local t = sdk.shared_data.get_variable_type("cs_result")

    if t == sdk.SharedDataType.String then
        imgui.text("C# wrote: " .. sdk.shared_data.get_variable_string("cs_result"))
    elseif t == sdk.SharedDataType.Number then
        imgui.text("C# wrote: " .. sdk.shared_data.get_variable_number("cs_result"))
    elseif t == sdk.SharedDataType.REManagedObject then
        local obj = sdk.shared_data.get_variable_managed_object("cs_result")
        imgui.text("C# wrote object: " .. obj:get_type_definition():get_full_name())
    else
        imgui.text("No data from C# yet")
    end
end)

File: csharp-api/test/Test/SharedData/TestInterop.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Dynamic;
using System.Reflection;
using Hexa.NET.ImGui;
using REFrameworkNET;
using REFrameworkNET.Callbacks;
using REFrameworkNET.Attributes;

public class MyPlugin {
    [Callback(typeof(ImGuiRender), CallbackType.Pre)]
    public static void RenderImGui() {
        // Read what Lua wrote
        var appType = SharedData.GetVariableType("app");
        var appTypeFullName = "None";
        if (appType == SharedDataType.REManagedObject) {
            var app = SharedData.GetVariableManagedObject("app");
            appTypeFullName = app.GetTypeDefinition().GetFullName();
            //API.LogInfo("Read via.Application from Lua via SharedData!");
        }

        var pi = SharedData.GetVariableNumber("pi");
        var count = SharedData.GetVariableNumber("count");
        var stringVal = SharedData.GetVariableString("stringVal");

        //API.LogInfo($"Lua sent pi={pi}, count={count}, stringVal={stringVal}");
        if (ImGui.Begin("TestInterop.cs")) {
            ImGui.Text($"Lua sent pi={pi}, count={count}, stringVal={stringVal}, appType={appTypeFullName}");
            ImGui.End();
        }

        // Write back for Lua to read
        SharedData.SetVariable("cs_result", pi * 2.0);
    }

    // public void OnFrame() {
    //     // You can also exchange data outside of draw
    //     if (SharedData.GetVariableType("app") != SharedDataType.None) {
    //         // Lua already set the data, do something
    //     }
    // }
}